### PR TITLE
Restrict HubProtocol on Server + HubOptions<THub>

### PR DESCRIPTION
**Pull request by mikaelm12**
Adding the ability to restrict the HubProtocol on the server
Issue: https://github.com/aspnet/SignalR/issues/854

### CODE REVIEW

**Comment by davidfowl** [comment]
I think this is way too global (I was going to update the issue but this is fine). It needs to be per hub end point not per signalr instance. We might need HubOptions<THub> to support this properly.

File: src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public static class DefaultProtocols
+    {
+        public static readonly string JSONProtocol = "json";
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs]: Use `Json`. We generally treat it like a "Word" in our code, even though it is technically an acronym ;)

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs]: I made it all lowercase to match the name property on the JsonHubProtocol https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs#L31
I can change that too if we want.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs]: The only default protocol we have is JSON really. MSGpack is an addon.

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Common/DefaultProtocols.cs]: Okay, I can remove that from here. Was following what the issue laid out 😄

**Comment by davidfowl** [comment]
We'll remove the HubOptions, in the AddSignalR overload and instead make them `HubOptions<THub>`. We'll revisit global options later.

`IHubProtocol.GetProtocol` now takes 3 arguments:

```C#
GetProtocol(IReadOnlyList<string> supportedProtocols,
string requestedProtocols,
HubConnectionContext context); 
```

We also need to add an extension method AddHubOptions on ISignalRBuilder to enable configuring this:

```C#
services.AddSignalR()
              .AddHubOptions<Chat>(o =>
              {
                   o.KeepAliveTimeout = TimeSpan.FromSeconds(5);
                   o.SupportedProtocols.Add("msgpack");
              });
```

**Comment by mikaelm12** [comment]
Okay, replaced `HubOptions` with `HubOptions<THub>`. HubOptions are no longer global. This probably could have logically been split into two PRs but oh well. 
The internals are quite different now btw

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs [hunk 2]
```diff
@@ -24,5 +25,7 @@ public class HubOptions
         /// functionality because of the polling mechanism. 
         /// </remarks>
         public TimeSpan KeepAliveInterval { get; set; } = DefaultKeepAliveInterval;
+
+        public List<string> SupportedProtocols { get; set; } = new List<string> { DefaultProtocols.JSONProtocol, DefaultProtocols.MsgPackProtocol};
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: I was waiting for this to come up. I think by default you support all available transports. That is, not specified means all. There also needs to be validation like what if you clear all? What does that mean?

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: Could we somehow grab all available transports when HubOptions is created and populate the list before running the users action on it?

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: We have nested generics so I'm not sure how well it will work. You'd need to get all the configured IHubProtocol implementations inside of an `IConfigureOptions<HubOptions<THub>>` which makes it a bit difficult as there's no explicit call to register a hub.

[the remark this answers is not in the document]
Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: Other fun things: MsgPack is in a completely separate assembly that isn't available in `AspNetCore.All` currently.

File: samples/SocketsSample/Startup.cs [hunk 3]
```diff
@@ -20,11 +20,25 @@ public void ConfigureServices(IServiceCollection services)
...
-            services.AddSignalR(options =>
+            services.AddSignalR()
+
+            .AddHubOptions<Chat>(options =>
+            {
+                options.KeepAliveInterval = TimeSpan.FromSeconds(5);
+            })
+            .AddHubOptions<DynamicChat>(options =>
             {
-                // Faster pings for testing
```
Comment (mikaelm12) [inline samples/SocketsSample/Startup.cs]: Don't know why I removed this. I'll bring it back

File: src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs [hunk 4]
```diff
@@ -15,10 +18,49 @@ public static ISignalRBuilder AddSignalR(this IServiceCollection services)
                 .AddJsonProtocol();
         }
 
+        public static ISignalRBuilder AddGlobalHubOptions(this ISignalRBuilder signalRBuilder, Action<HubOptions> options)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs]: This should be an overload of AddSignalR. There should be no global options call.

File: src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs [hunk 5]
```diff
@@ -15,10 +18,49 @@ public static ISignalRBuilder AddSignalR(this IServiceCollection services)
...
+        }
+
         public static ISignalRBuilder AddHubOptions<THub>(this ISignalRBuilder signalrBuilder, Action<HubOptions<THub>> options) where THub : Hub
         {
             signalrBuilder.Services.Configure(options);
             return signalrBuilder;
         }
     }
+
+    public class HubOptionsSetup : IConfigureOptions<HubOptions>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs]: Put these in separate files.

File: src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs (hunk 4 above)
Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs]: Makes sense

**Comment by mikaelm12** [comment]
Okay, this is working now and is ready for review

File: samples/SocketsSample/Startup.cs [hunk 6]
```diff
@@ -20,11 +20,11 @@ public void ConfigureServices(IServiceCollection services)
         {
             services.AddSockets();
 
-            services.AddSignalR(options =>
-            {
+            services.AddSignalR(options => {
                 // Faster pings for testing
                 options.KeepAliveInterval = TimeSpan.FromSeconds(5);
             })
+
```
Comment (davidfowl) [inline samples/SocketsSample/Startup.cs]: nit: remove space

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            foreach (IHubProtocol hubProtocol in protocols)
+            {
+                _protocols.Add(hubProtocol.Name);
+            }
+        }
+
+        public void Configure(HubOptions options)
+        {
+            if( options.SupportedProtocols == null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs]: nit: formatting

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup`T.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        private readonly HubOptions _hubOptions;
+        public HubOptionsSetup(IOptions<HubOptions> options)
+        {
+            _hubOptions = options.Value;
+        }
+
+        public void Configure(HubOptions<THub> options)
+        {
+            if(_hubOptions.SupportedProtocols == null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup`T.cs]: Can this ever be null?

File: src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs [hunk 9]
```diff
@@ -33,11 +31,11 @@ public DefaultHubProtocolResolver(IOptions<HubOptions> options, IEnumerable<IHub
             }
         }
 
-        public IHubProtocol GetProtocol(string protocolName, HubConnectionContext connection)
+        public IHubProtocol GetProtocol(string protocolName, List<string> supportedProtocols, HubConnectionContext connection)
         {
             protocolName = protocolName ?? throw new ArgumentNullException(nameof(protocolName));
 
-            if (_availableProtocols.TryGetValue(protocolName, out var protocol))
+            if (_availableProtocols.TryGetValue(protocolName, out var protocol) && (supportedProtocols ==  null || supportedProtocols.Contains(protocolName)))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs]: What about casing? Do the case rules match the dictionary? It should be insensitive.

File: src/Microsoft.AspNetCore.SignalR.Core/SignalRDependencyInjectionExtensions.cs [hunk 10]
```diff
@@ -4,6 +4,7 @@
 using Microsoft.AspNetCore.SignalR;
 using Microsoft.AspNetCore.SignalR.Core;
 using Microsoft.AspNetCore.SignalR.Internal;
+using Microsoft.Extensions.Options;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/SignalRDependencyInjectionExtensions.cs]: Revert this file?

File: src/Microsoft.AspNetCore.SignalR/HubOptionsDependencyInjectionExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+using System;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class HubOptionsDependencyInjectionExtensions
+    {
+
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubOptionsDependencyInjectionExtensions.cs]: nit: remove empty space.

File: src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs [hunk 12]
```diff
@@ -3,22 +3,25 @@
...
-        {
-            services.Configure(configure);
             services.AddSockets();
             return services.AddSignalRCore()
                 .AddJsonProtocol();
         }
+
+        public static ISignalRBuilder AddSignalR(this IServiceCollection services, Action<HubOptions> options)
+        {
+            services.AddSingleton<IConfigureOptions<HubOptions>, HubOptionsSetup>()
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/SignalRDependencyInjectionExtensions.cs]: This should be in the parameterless AddSignalR

File: test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs [hunk 13]
```diff
@@ -274,7 +274,7 @@ public class HubEndpointTests
         {
             var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
             {
-                services.Configure<HubOptions>(hubOptions =>
+                services.Configure<HubOptions<SimpleHub>>(hubOptions =>
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs]: Why isn't this just the same as before? why did you change it?

File: test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs [hunk 14]
```diff
@@ -1325,10 +1325,15 @@ public void HubsCannotHaveOverloadedMethods()
         [MemberData(nameof(StreamingMethodAndHubProtocols))]
         public async Task HubsCanStreamResponses(string method, IHubProtocol protocol)
         {
-            var serviceProvider = HubEndPointTestUtils.CreateServiceProvider();
+            var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
+            {
+                services.AddSignalR(options =>
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs]: Why did you add this?

File: test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs [hunk 15]
```diff
@@ -1673,7 +1678,7 @@ public static IEnumerable<object[]> StreamingMethodAndHubProtocols
         public async Task DoesNotWritePingMessagesIfSufficientOtherMessagesAreSent()
         {
             var serviceProvider = HubEndPointTestUtils.CreateServiceProvider(services =>
-                services.Configure<HubOptions>(options =>
+                services.Configure<HubOptions<MethodHub>>(options =>
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs]: Why did you change these?

File: test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs [hunk 16]
```diff
@@ -22,18 +22,48 @@ public class DefaultHubProtocolResolverTests
             new MessagePackHubProtocol()
         };
 
+        private static readonly List<string> AllProtocolNames = new List<string> { "json", "messagepack" };
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs]: Why do you have this?

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup`T.cs (hunk 8 above)
Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup`T.cs]: It shouldn't be. I'll remove this check

**Comment by mikaelm12** [comment]
Okay. The tests weren't actually testing the correct thing before. I fixed them to ensure that the global hub options are being configured.

I changed the HubOptions<THub> class values to be null by default to add logic to determine if we should use the global options. We have to do this because configure on HubOptionsSetup<THub> doesn't run unless we explicitly call `addHubOptions` for a specific THub.

File: samples/SocketsSample/Startup.cs [hunk 17]
```diff
@@ -20,8 +20,7 @@ public void ConfigureServices(IServiceCollection services)
         {
             services.AddSockets();
 
-            services.AddSignalR(options =>
-            {
+            services.AddSignalR(options => {
```
Comment (BrennanConroy) [inline samples/SocketsSample/Startup.cs]: revert

File: src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs [hunk 18]
```diff
@@ -32,16 +34,30 @@ public class HubEndPoint<THub> where THub : Hub
...
+            _globalHubOptions = globalHubOptions.Value;
             _logger = loggerFactory.CreateLogger<HubEndPoint<THub>>();
             _userIdProvider = userIdProvider;
             _dispatcher = dispatcher;
         }
 
         public async Task OnConnectedAsync(ConnectionContext connection)
         {
-            var connectionContext = new HubConnectionContext(connection, _hubOptions.KeepAliveInterval, _loggerFactory);
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: Add a comment saying `HubOptions<T>` is preferred over `HubOptions`

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class HubOptionsSetup : IConfigureOptions<HubOptions>
+    {
+        private readonly List<string> _protocols = new List<string>();
+
+        public HubOptionsSetup(IEnumerable<IHubProtocol> protocols)
+        {
+            foreach (IHubProtocol hubProtocol in protocols)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs]: var

File: src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs [hunk 20]
```diff
@@ -33,11 +32,11 @@ public DefaultHubProtocolResolver(IOptions<HubOptions> options, IEnumerable<IHub
             }
         }
 
-        public IHubProtocol GetProtocol(string protocolName, HubConnectionContext connection)
+        public IHubProtocol GetProtocol(string protocolName, List<string> supportedProtocols, HubConnectionContext connection)
         {
             protocolName = protocolName ?? throw new ArgumentNullException(nameof(protocolName));
 
-            if (_availableProtocols.TryGetValue(protocolName, out var protocol))
+            if (_availableProtocols.TryGetValue(protocolName, out var protocol) && (supportedProtocols ==  null || supportedProtocols.Contains(protocolName, StringComparer.OrdinalIgnoreCase)))
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Core/Internal/DefaultHubProtocolResolver.cs]: nit: remove extra space

File: src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs [hunk 21]
```diff
@@ -2,11 +2,12 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+using System.Collections.Generic;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs]: nit: sort

File: test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs [hunk 22]
```diff
@@ -22,18 +22,47 @@ public class DefaultHubProtocolResolverTests
...
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs]: Why are you hardcode this?

Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs]: move to top of the file?

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class HubOptions<THub> : HubOptions where THub : Hub
+    {
+        public new TimeSpan? NegotiateTimeout { get; set; } = null;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs]: It feels really weird to me that these properties are defined again and are using new to hide the original properties. Is it because you want NegotiateTimeout and KeepAliveInterval to be nullable here? What scenario is this for?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs]: Yea, maybe they should just be null in the base class.

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs]: Good question. 
The properties are null in this class and not the base class because we want the default values from the base class, but we need to be able to tell if a specific `HubOptions<THub>` has been configured (aka the options are not null) 
It allows us to differentiate between the scenarios in which there are only global options defined, global and local so we know which settings to enforce. 

That being said, yeah we could just make them null in the base class but I do like the idea of default values.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions`T.cs]: We can just have defaults come from the IConfigureOptions<HubOptions> instead of being on the HubOptions class. It's a bit cleaner.

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs [hunk 24]
```diff
@@ -26,6 +27,18 @@ public void Configure(HubOptions options)
...
             }
+
+            if (options.KeepAliveInterval == null)
+            {
+                // The default keep - alive interval.This is set to exactly half of the default client timeout window,
+                // to ensure a ping can arrive in time to satisfy the client timeout.
+                options.KeepAliveInterval = TimeSpan.FromSeconds(15);
+            }
+
+            if(options.NegotiateTimeout == null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptionsSetup.cs]: nit: formatting

File: src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs [hunk 25]
```diff
@@ -32,16 +34,31 @@ public class HubEndPoint<THub> where THub : Hub
...
+
+            if (supportedProtocols != null && supportedProtocols.Count == 0)
+            {
+                throw new InvalidOperationException("There are no supported protocols");
+            }
+
+            var connectionContext = new HubConnectionContext(connection, keepAlive.Value, _loggerFactory);
 
-            if (!await connectionContext.NegotiateAsync(_hubOptions.NegotiateTimeout, _protocolResolver, _userIdProvider))
+            if (!await connectionContext.NegotiateAsync(negotiateTimeout.Value, supportedProtocols, _protocolResolver, _userIdProvider))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: This can throw if negotiateTimeout is null...

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: It shouldn't be null since it gets set in the HubOptionsSetup

Comment (davidfowl) [reply src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: What if I explicitly set it to null?

File: src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs [hunk 26]
```diff
@@ -32,16 +34,31 @@ public class HubEndPoint<THub> where THub : Hub
...
             _userIdProvider = userIdProvider;
             _dispatcher = dispatcher;
         }
 
         public async Task OnConnectedAsync(ConnectionContext connection)
         {
-            var connectionContext = new HubConnectionContext(connection, _hubOptions.KeepAliveInterval, _loggerFactory);
+            // We check to see if HubOptions<THub> are set because those take precedence over global hub options.
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval;
+            var negotiateTimeout = _hubOptions.NegotiateTimeout ?? _globalHubOptions.NegotiateTimeout;
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: I think this makes sense.

File: src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs [hunk 27]
```diff
@@ -32,16 +34,32 @@ public class HubEndPoint<THub> where THub : Hub
...
 
         public async Task OnConnectedAsync(ConnectionContext connection)
         {
-            var connectionContext = new HubConnectionContext(connection, _hubOptions.KeepAliveInterval, _loggerFactory);
+            // We check to see if HubOptions<THub> are set because those take precedence over global hub options.
+            // Then set the keepAlive and negotiateTimeout values to the defaults in HubOptionsSetup incase they were explicitly set to null.
+            var keepAlive = _hubOptions.KeepAliveInterval ?? _globalHubOptions.KeepAliveInterval ?? HubOptionsSetup.DefaultKeepAliveInterval;
+            var negotiateTimeout = _hubOptions.NegotiateTimeout ?? _globalHubOptions.NegotiateTimeout ?? HubOptionsSetup.DefaultNegotiateTimeout;
+            var supportedProtocols = _hubOptions.SupportedProtocols;
+            if (supportedProtocols == null && _globalHubOptions.SupportedProtocols != null)
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs]: You could replace this if statement with a null coalesce on the line above. It doesn't matter if `_globalHubOptions.SupportedProtocols` protocols is null based on the check line 55.

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs [hunk 28]
```diff
@@ -2,27 +2,16 @@
...
     {
-        /// <summary>
-        /// The default keep-alive interval. This is set to exactly half of the default client timeout window,
-        /// to ensure a ping can arrive in time to satisfy the client timeout.
-        /// </summary>
-        public static readonly TimeSpan DefaultKeepAliveInterval = TimeSpan.FromSeconds(15);
+        public TimeSpan? NegotiateTimeout { get; set; } = null;
 
-        public TimeSpan NegotiateTimeout { get; set; } = TimeSpan.FromSeconds(5);
+        public TimeSpan? KeepAliveInterval { get; set; } = null;
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: Bring the docs back, and add a note about `null` indicating the default interval (15 seconds).

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs [hunk 29]
```diff
@@ -2,27 +2,16 @@
...
-        /// <summary>
-        /// The interval at which keep-alive messages should be sent. The default interval
-        /// is 15 seconds.
-        /// </summary>
-        /// <remarks>
-        /// This interval is not used by the Long Polling transport as it has inherent keep-alive
-        /// functionality because of the polling mechanism. 
-        /// </remarks>
-        public TimeSpan KeepAliveInterval { get; set; } = DefaultKeepAliveInterval;
+        public List<string> SupportedProtocols { get; set; } = null;
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: Make this `IList<string>`

We're saying `null` is "any protocol" right? Let's throw a doc comment on this explaining it as well.

File: src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs [hunk 30]
```diff
@@ -1,12 +1,13 @@
...
 
+using System.Collections.Generic;
 using Microsoft.AspNetCore.SignalR.Internal.Protocol;
 
 namespace Microsoft.AspNetCore.SignalR.Internal
 {
     public interface IHubProtocolResolver
     {
-        IHubProtocol GetProtocol(string protocolName, HubConnectionContext connection);
+        IHubProtocol GetProtocol(string protocolName, List<string> supportedProtocols, HubConnectionContext connection);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs]: `List<T>` in a public API!



Use `IReadOnlyList<string>`. The method only needs to be able to read from the list, and doesn't care what kind of list you give it.

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs]: This should also just be an `IList<string>`, right? It get passed the supportedProtocol list from the HubEndPoint which is an IList

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs (hunk 28 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: The comment is a bit more complicated than that because `null` could indicate the default, or it could indicate that the global option should be used

File: src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs (hunk 30 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs]: Should it not be `List<T>` even if it is in the internal namespace? I wasn't sure about that when I looked at the code.

Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Core/Internal/IHubProtocolResolver.cs]: In this case, the API is likely to be made properly public at some point. Also, "pubternal" (public in the Internal namespace) types should still try to keep to our public API standards.

File: src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs (hunk 28 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs]: Ugh, yeah. Ok, save it for now, we'll need to doc it at some point though.
